Organize JavaScript-externals manage page dependencies

moorscode opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* Reduce the size of the JavaScript that is being loaded on page where it is not needed.
* Goals: quicker page-loads & less data transferred per page

### Todo

**Test with the addons**
* [ ] Premium - [PR](https://github.com/Yoast/wordpress-seo-premium/pull/2978)
* [ ] Local - only uses `yoastseo: window.yoast.analysis` as external - no adjustments needed
* [ ] WooCommerce - only uses `yoastseo: window.yoast.analysis` as external - no adjustments needed
* Video - doesn't use webpack
* News - doesn't use webpack

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another repo, start you changelog item with the repo name between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/repos, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Improved the loading times of admin-pages where Yoast SEO places content.

## Relevant technical choices:

* Introduced 10 additional JavaScript-files to be loaded; though this helps to reduce the number of files that need to be loaded on *most* pages. The post-edit page already has ~60 files that are loaded (WordPress loads a lot of small files); adding 10 there did not seem like a big impact.
* Specified -all- dependencies for all assets declared in our admin-asset-manager (a lot were missing)
* Created a `components` which loads the dependencies that are currently tied to the `components-legacy` (yoast-components) asset
* Created a `components-modern` which is the `@yoast/components` package
* _More in the JavaScript repository-branch_

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
This PR can be tested by following these steps:

* Visit all the pages we place javascript on and don't see any JavaScript errors and see that everything is loading/working as expected.
* WordPress Dashboard
* Post/term overview
* Post/term edit
* Settings pages
* Configuration wizard

## UI changes
* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation
* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes PROD-202
Requires https://github.com/Yoast/javascript/pull/779

omarreiss commented on admin/class-admin-asset-manager.php:
```diff
@@ -460,6 +549,90 @@ protected function scripts_to_be_registered() {
 			],
+			[
+				'name' => 'helpers',
+				'src'  => 'yoast-helpers-' . $flat_version,
```
Please don't prefix filenames. Add an extra directory instead.

moorscode replied:
Done

omarreiss commented on webpack/webpack.config.js:
```diff
@@ -68,6 +68,19 @@ const yoastExternals = {
 	"@yoast/style-guide": "window.yoast.styleGuide",
 };
 
+const yoastExternalsMapping = {
```
> What's the difference between `yoastExternals` and `yoastExternalsMapping`? 
...is what everyone will ask. Which is a clear signal the solution that follows is probably not simple enough.

herregroen commented on webpack/webpack.config.js:
```diff
@@ -117,6 +130,38 @@ function createBundleAnalyzer() {
 	} );
 }
 
+function addYoastExternals( config, base, plugins, pluginVersionSlug ) {
```
Looking a bit further at how Gutenberg handles this I think we can do this much more elegantly.

If we use a webpack config with the library option as WP uses: https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L73

We can point the entries directly at the node_modules entry files and don't have to include any local scripts nor handle assigning them to the window ourselves. That'd remove this complexity of generating webpack configs and remove all those files from the externals directory.

omarreiss replied:
Ah yes, this would be even more preferable.

omarreiss commented on webpack/webpack.config.js:
```diff
@@ -117,6 +130,38 @@ function createBundleAnalyzer() {
 	} );
 }
 
+function addYoastExternals( config, base, plugins, pluginVersionSlug ) {
```
This works, but for 99% of the config that is duplicated here dynamically it's impossible to follow _why_. Only when you understand our particular setup with externals is it possible to understand this config. I'd rather see this solved differently. Without this, the confusing object `YoastExternalsMapping` is also not needed anymore. Please see if you can use the [dependency extraction webpack plugin](https://github.com/WordPress/gutenberg/tree/master/packages/dependency-extraction-webpack-plugin) for this.  If it doesn't support configuration yet, it might be easy to contribute that back so we don't have to fork the entire plugin.

moorscode commented:
For personal context: https://github.com/WordPress/gutenberg/blob/master/packages/dependency-extraction-webpack-plugin/lib/util.js

moorscode commented on admin/class-admin-asset-manager.php:
```diff
@@ -431,21 +474,67 @@ protected function scripts_to_be_registered() {
+				 */
 				'name' => 'components',
-				'src'  => 'components-' . $flat_version,
+				'src'  => false,
```
Done

moorscode commented:
I could not see how we could use the package from Gutenberg, as that mostly generates PHP files that can be used to register assets. That would require reworking the asset-manager, which is too far out of scope for this focus.

I have simplified the webpack configuration and extracted the externals to a separate configuration file.
The confusing variable has been removed, we are only talking about packages and externals in there.

The additional function has also been removed, by just doing it inline. Which makes it a lot easier to follow and keep it in context of the other configuration parts.

moorscode commented:
The configuration wizard and settings bundles in the `dist` folder both have `react-select` built-in. I've tried to create an external of that but I couldn't seem to get that working.

I'm expecting that to be a problem because `react-select` is only used in the external `@yoast/configuration-wizard` package and might be included because of it.. not confident though that is actually the problem.

This code is around `210kb` "stat" size in the bundle analyzer, so could safe a little space.

@herregroen do you have any insights?

omarreiss commented on admin/class-admin-asset-manager.php:
```diff
@@ -452,21 +513,67 @@ protected function scripts_to_be_registered() {
+			],
+			[
+				// The `@yoast/components` package.
+				'name' => 'components-modern',
```
If we have `components-legacy`, this can just be `components`

moorscode replied:
We currently have a `components` asset; which contains all the packages.
I didn't think it would be wise to just replace the contents of the asset with a small subset of what it currently is.

Following your suggestion to rename the legacy components, I suggest we use `yoast-components` for this asset.

omarreiss commented on admin/class-admin-asset-manager.php:
```diff
@@ -452,21 +513,67 @@ protected function scripts_to_be_registered() {
+			],
+			[
+				// The `yoast-components` package.
+				'name' => 'components-legacy',
```
Let's call this `legacy-components`

moorscode commented:
> What I didn't expect is to see so many changes in JavaScript files themselves. Why is that needed? The dependencies shouldn't have changed on the file level right? When testing with develop monorepo, there's not a single interface that's not broken.

I needed to be able to access some functions without loading the whole `yoastseo` package on specific pages.
This was present on the search-appearance/settings pages where we show the replacement variable editor.
This should not require `yoastseo` to be loaded in to work.

The specifics on which helps are moved are mentioned on the PR on the JavaScript repository; https://github.com/Yoast/javascript/pull/779

The JavaScript code changes in this PR are only applying the moves/duplications that are done there.

I'm not super happy with duplicating stuff, but it seemed the best step to move forward and solve the duplication problem in a next iteration. Also discussed with @herregroen.
